### PR TITLE
Pass context RouteData to ActionContext before execution

### LINKED ISSUE #4267: Returning a AcceptedAtRouteResult throws "System.Private.CoreLib: Index was out of range." exception
A customer created a question on Stack Overflow that trying to return an AcceptedAtRouteResult from their function was throwing a "Index was out of range" exception in System.Private.CoreLib.

The code they provided does reproduce the issue on my local and in Azure.

You can pull the metrics from:
- Timestamp: 2019-04-04 21:13:10.025
- Function App version (1.0 or 2.0): 2.0, Microsoft.NET.Sdk.Functions 1.0.26
- Function App name: sasoucyfunctionapp
- Function name(s) (as appropriate): AcceptedAtRouteResult
- Invocation ID: eff6a9ac-17e4-4dd5-a3bf-e5615fab45a0
- Region: East US 2

#### Expected behavior

Function should return 200 status with the set parameters

#### Actual behavior

Function completes successfully but then returns a 500 error

Provide any related information 

- Language: C#
- Source:  https://stackoverflow.com/questions/55453819/how-to-return-acceptedatrouteresult-from-an-azure-function
<details>
<summary>Source</summary>

```csharp
[FunctionName("AcceptedAtRouteResult")]
public static IActionResult AcceptedAtRouteResult(
    [HttpTrigger("GET")]HttpRequest req)
{
    // read query parameter if present else set to defualt value

   var rs = new AcceptedAtRouteResult(
        "acceptedatrouteresult", 
        new { someParameter = "value" }, 
        new { Result = "1" });

    return rs;
}
```
</details>

### PR DESCRIPTION
**Pull request by brandonh-msft**
Fixes #4267

### CODE REVIEW

**Comment by brandonh-msft** [comment]
cc @brettsam & @fabiocav for this fix. See associated issue for RCA & justification of change

**Comment by fabiocav** [comment]
@brandonh-msft are you intending to submit any tests with this?

**Comment by brandonh-msft** [comment]
I explored the idea locally, but wasn't able to make heads or tails of how to get a function to execute **and** run all the way through the middleware pipeline to where this code would be hit. But I'm happy to work with Brett or you tomorrow on getting a test in as part of this PR if you like. Thanks for taking a look!

**Comment by brandonh-msft** [comment]
working on these now!

**Comment by brandonh-msft** [comment]
@fabiocav :)
as you wish

cc @brettsam

**Comment by brandonh-msft** [comment]
⏲

**Comment by fabiocav** [comment]
Target branch has been updated, we can merge and backport this.

**Comment by fabiocav** [comment]
Thanks for updating. One last ask, could you please squash this to a single commit so we can merge? Thanks again @brandonh-msft

**Comment by brandonh-msft** [comment]
@fabiocav :)
as you wish

**Comment by cecilphillip** [comment]
@fabiocav @brandonh-msft curious where this landed.

**Comment by brandonh-msft** [comment]
> @fabiocav @brandonh-msft curious where this landed.

Thanks for the ping, @cecilphillip. From what I can tell this PR just died; now I have some conflict resolution to tend to. I'll attempt to work on this today and push an update to the branch for the PR to continue.

**Comment by brandonh-msft** [comment]
@fabiocav / @brettsam I had a real issue trying to get the UTs to run this time around. I had the `AzureWebJobsStorage` env var set in both my system env and the Project env settings but every time tests spun up they complained they couldn't find a value for it and died. Setting a breakpoint on the blowup and inspecting the config providers shows the value present, though. I think our Contributor docs need updating in this regard because I can only assume they're missing some detail here.

That said - assuming server CIs end up being happy w/ the change (I've waited quite a while and they still haven't run) - ⏲️ to y'all again.

Thanks again @cecilphillip for the poke on this :)

File: src/WebJobs.Script.WebHost/Middleware/FunctionInvocationMiddleware.cs [hunk 1]
```diff
@@ -61,7 +61,7 @@ public async Task Invoke(HttpContext context)
                     return;
                 }
 
-                ActionContext actionContext = new ActionContext(context, _emptyRouteData, _emptyActionDescriptor);
+                ActionContext actionContext = new ActionContext(context, context.GetRouteData(), new ActionDescriptor());
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Middleware/FunctionInvocationMiddleware.cs:56]: Remove the (now) unused variables `_emptyRouteData` and `_emptyActionDescriptor`, though for the descriptor, we might want to just continue using the previously created/allocated one.

**Comment by fabiocav** [comment]
Looking good, minor comment, but yeah, let's get this in.

**Comment by brandonh-msft** [comment]
good catch, thanks. cleaned up.

**Review by fabiocav — APPROVED** [review]
@brandonh-msft this looks good!

Would you mind squashing the commits so we can merge this in (rebase and squash would be ideal)

**Comment by fabiocav** [comment]
Merged

**Pull request merged**